### PR TITLE
feat(partition): add legend and highlighters

### LINKED ISSUE #486: Decouple legend and tooltip from XYChart
**Is your feature request related to a problem? Please describe.**
The current implementation of Legend and the tooltip is partially coupled with XYChart type.
We should decouple it to allow reusing the Legend and the tooltip freely among all chart types.

**Describe the solution you'd like**
Move Legend and Tooltip into the global component folder, decouple LegendItems and associated types from XYChart.

- [x] Tooltip
- [x] Legend

### LINKED ISSUE #532: Partition charts with legend and tooltip
Legends and tooltips are a requirement for replacing pie and donut charts in Kibana. The main reason is that when rendered in a dashboard, it's not obvious what the visualization shows without the tooltip to indicate the source of the data.

The legend and coloring implementation that exists in Kibana combines all values into a flat legend, regardless of the layer it's displayed on. That behavior might not be needed if we can come up with a better way of displaying legends.

Screenshot 2020-01-30 14 27 50
Screenshot 2020-01-30 14 28 42

While there are customization options in the current implementation, they are pretty minimal. For example, the legend can't be hidden.
Screenshot 2020-01-30 14 29 46

### PR DESCRIPTION
**Pull request by markov00**
## Summary

This PR adds the legend to pie charts and treemaps.
Close #486 finishing the decoupling of the legend from XY charts
Close #532 request to add legend and tooltips to partition charts

Screenshot 2020-04-09 at 13 06 04

Two more options are added to the `<Settings/>` components
- `flatLegend` will flat the legend for pie/tree hierarchical legend as in the following screenshot
Screenshot 2020-04-09 at 13 06 23

- `legendMaxDepth` will limit the legend hierarchy to a maximum depth value: `1` is the root level. Can be combined with the `flatLegend`

`legendMaxDepth={1}` example on a 3 level piechart
Screenshot 2020-04-09 at 13 06 38

`legendMaxDepth={2}` example on a 3 level piechart
Screenshot 2020-04-09 at 13 06 52

The PR also add the highlighter when hovering over the pie/tree chart and over the legend items as shown in the following gifs
pielight
piedark
treedark
treelight

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [x] Unit tests were updated or added to match the most common scenarios

### CODE REVIEW

**Comment by wylieconlon** [comment]
Based on the screenshots and gifs, this looks like exactly what we are looking for! Thank you!

File: stories/legend/10_sunburst.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+export const example = () => {
+  const flatLegend = boolean('flatLegend', true);
+  const legendMaxDepth = number('legendMaxDepth', 2, {
+    min: 0,
+    max: 3,
+    step: 1,
+  });
+
+  return (
+    <Chart className="story-chart" /*size={{ width: 1200, height: 800 }}*/>
```
Comment (rshen91) [inline stories/legend/10_sunburst.tsx:42]: You might want to remove this comment?

File: src/chart_types/partition_chart/renderer/dom/highlighter.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+ * @param r The arc's radius. Must be positive
+ * @param a0 The angle at which the arc starts in radians, measured from the positive x-axis
+ * @param a1 The angle at which the arc ends in radians, measured from the positive x-axis
+ * @param ccw If 1, draws the arc counter-clockwise between the start and end angles
+ */
+function getSectorShapeFromCanvasArc(x: number, y: number, r: number, a0: number, a1: number, ccw: number): string {
+  const cw = 1 ^ ccw;
+  const da = ccw ? a0 - a1 : a1 - a0;
+  return `A${r},${r},0,${+(da >= Math.PI)},${cw},${x + r * Math.cos(a1)},${y + r * Math.sin(a1)}`;
+}
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/partition_chart/renderer/dom/highlighter.tsx:64]: fixed in d5539128a83524c7ee6582ab47b2f6a876008853

File: src/chart_types/partition_chart/renderer/dom/highlighter.tsx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+ * Partition chart highlighter from legend events
+ * @internal
+ */
+export const HighlighterFromLegend = connect(legendMapStateToProps)(HighlighterComponent);
+
+/**
+ * Partition chart highlighter from mouse hover events
+ * @internal
+ */
+export const HighlighterFromHover = connect(hoverMapStateToProps)(HighlighterComponent);
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/partition_chart/renderer/dom/highlighter.tsx:273]: fixed in 7c28da1c4cf3a89f40f3083dd86ba9ba92ba6a86

File: src/components/legend/legend_item.tsx [hunk 4]
```diff
@@ -176,28 +142,37 @@ export class LegendListItem extends Component<LegendItemProps, LegendItemState>
...
           onMouseEnter={this.onLegendItemMouseOver}
           onMouseLeave={this.onLegendItemMouseOut}
+          style={style}
         >
-          {this.renderColor(color, isSeriesVisible)}
-          {renderLabel(onLabelClick, hasLabelClickListener, label)}
-          {showExtra && renderExtra(extra, isSeriesVisible)}
-        </div>
+          {renderColor(color, isSeriesHidden, hasColorPicker, colorClick)}
+          {renderLabel(label, onLabelClick, hasLabelClickListener)}
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/legend/legend_item.tsx:173]: done in 8119703600a65a0ae136f281377e137dce78f5e3

File: src/components/legend/legend_item.tsx [hunk 5]
```diff
@@ -19,69 +19,78 @@
...
+  position: Position;
+  extraValues: Map<string, LegendItemExtraValues>;
   showExtra: boolean;
-  legendColorPicker?: LegendColorPicker;
+  colorPicker?: LegendColorPicker;
   onLegendItemClickListener?: LegendItemListener;
   onLegendItemOutListener?: BasicListener;
   onLegendItemOverListener?: LegendItemListener;
   legendItemOutAction: typeof onLegendItemOutAction;
   legendItemOverAction: typeof onLegendItemOverAction;
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/legend/legend_item.tsx:52]: done in 8119703600a65a0ae136f281377e137dce78f5e3

File: src/chart_types/partition_chart/state/selectors/compute_legend.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+          },
+        };
+      });
+  },
+)(getChartIdSelector);
+
+function findIndex(items: Array<[PrimitiveValue, number, PrimitiveValue]>, child: QuadViewModel) {
+  return items.findIndex((item) => {
+    return item[0] === child.dataName && item[1] === child.depth && item[2] === child.value;
+  });
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/partition_chart/state/selectors/compute_legend.ts:100]: done in 8119703600a65a0ae136f281377e137dce78f5e3

File: src/chart_types/partition_chart/state/selectors/compute_legend.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        excluded.add(key);
+      }
+      return true;
+    });
+
+    if (forceFlatLegend) {
+      items = items.sort((a, b) => {
+        return a.depth - b.depth;
+      });
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/partition_chart/state/selectors/compute_legend.ts:69]: done in 8119703600a65a0ae136f281377e137dce78f5e3

**Comment by markov00** [comment]
jenkins test this

File: src/chart_types/partition_chart/state/selectors/compute_legend.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    const uniqueNames = geoms.quadViewModel.reduce<Record<string, number>>((acc, { dataName, fillColor }) => {
+      const key = [dataName, fillColor].join('---');
+      if (!acc[key]) {
+        acc[key] = 0;
+      }
+      acc[key] += 1;
+      return acc;
+    }, {});
+
+    const { flatLegend, legendMaxDepth, legendPosition } = settings;
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/chart_types/partition_chart/state/selectors/compute_legend.ts:48]: Fixed by 11f120f17a82175d0f346e456fc021e81c74a4b5 👍

**Pull request merged**